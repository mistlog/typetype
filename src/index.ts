export * from "./parser";
export * from "./generator";
export * from "./core";