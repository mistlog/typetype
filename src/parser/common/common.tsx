import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}