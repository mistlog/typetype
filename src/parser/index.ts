export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";