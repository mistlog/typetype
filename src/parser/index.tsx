export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";