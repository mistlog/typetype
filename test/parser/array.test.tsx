import { ReactPeg } from "react-peg";
import { ArrayType, OperatorType, TypeExpression } from "../../src";
import { saveAST } from "../common";

describe("parser: array", () => {
    test("ArrayType", () => {
        const parser = ReactPeg.render(<ArrayType />);
        const ast = parser.parse(`string[]`);
        saveAST(ast, "ArrayType.json");
        expect(ast).toMatchSnapshot();
    })

    test("ArrayType: readonly", () => {
        const parser = ReactPeg.render(<OperatorType />);
        const ast = parser.parse(`readonly string[]`);
        saveAST(ast, "ArrayType-Readonly.json");
        expect(ast).toMatchSnapshot();
    })

    test("ArrayType: any", () => {
        const parser = ReactPeg.render(<ArrayType />);
        const ast = parser.parse(`any[]`);
        saveAST(ast, "ArrayType-Any.json");
        expect(ast).toMatchSnapshot();
    })

    test("ArrayType: deep", () => {
        const parser = ReactPeg.render(<TypeExpression />);
        const ast = parser.parse(`string[][]`);
        saveAST(ast, "ArrayType-Deep.json");
        expect(ast).toMatchSnapshot();
    })
})