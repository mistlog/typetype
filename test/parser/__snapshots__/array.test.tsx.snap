// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: array ArrayType 1`] = `
Object {
  "dimension": 1,
  "elementType": Object {
    "kind": "StringType",
    "value": "string",
  },
  "kind": "ArrayType",
}
`;

exports[`parser: array ArrayType: any 1`] = `
Object {
  "dimension": 1,
  "elementType": Object {
    "kind": "AnyType",
    "value": "any",
  },
  "kind": "ArrayType",
}
`;

exports[`parser: array ArrayType: deep 1`] = `
Object {
  "dimension": 2,
  "elementType": Object {
    "kind": "StringType",
    "value": "string",
  },
  "kind": "ArrayType",
}
`;

exports[`parser: array ArrayType: readonly 1`] = `
Object {
  "kind": "ReadonlyArray",
  "operand": Object {
    "dimension": 1,
    "elementType": Object {
      "kind": "StringType",
      "value": "string",
    },
    "kind": "ArrayType",
  },
}
`;
